### ISSUE TITLE
微调 BERT 实现自然语言意图识别

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#171]
微调 (Fine-tuning) BERT 系列模型，即在特定任务数据集上训练 BERT 使其适应特定任务，是一种主流且有效的自然语言处理 (NLP) 方法。BERT 及其变种（如 RoBERTa, ELECTRA, ALBERT 等）通过微调几乎可以应用于所有的主流 NLP 任务，并且在很多任务上取得了 **SOTA (State-Of-The-Art)** 的效果。

其应用场景非常广泛，主要可以分为：

1. 单文本分类（Single Text Classification），如 意图识别
2. 文本对分类/关系判断（Text Pair Classification/Relationship Judgement），如 rerank 模型
3. 文本生成（Text Generation）

### 1. 为什么做（Why）

微调的成本不低，所以在开始前最好谨慎思考如下问题：[^1]

1. 现有模型是不是真的无法满足你的需求
2. 是否有足够的专有领域、且高质量的数据
3. 是否有评估方法，可以确保微调有效，达到了预定效果

### 2. 要达到什么目标（What）

从实际需求出发，看要达到什么样的目标 🎯。例如在哪个领域达到多少准确率，是否考虑微调后在其他领域的退化等。

### 3. 要怎么做

有一些常见技术，可用在不同阶段，达成不同目标，要考虑如何选择来达到上述目标。例如：

1. Supervised Fine-tuning(SFT) ：监督微调，建立核心能力；
2. Preference optimization(PO) : 偏好优化，直接从人类或AI的偏好学习；
3. Reinforcement learning(RL) : 强化学习，可以超越监督数据的稳定性和推理能力；
4. Data curation : 平衡数据多样性和质量；
5. Evaluation : 跟踪进展，并及早发现倒退。

### 模型微调（训练）

使用准备好的数据和任务模型进行训练。

1. 设置优化器和学习率：通常使用 Adam W 优化器，并采用一个非常小的学习率（如 1e-5 到 5e-5），因为 BERT 模型参数已经经过大量预训练，过大的学习率可能会破坏其学到的知识。同时，通常采用 Warmup 和 Decay 的学习率调度策略。学习率的作用和设置参见 #167 
2. 设置评估指标：评估指标作用是告诉 Trainer：到底用哪个评估指标来判定“哪个 checkpoint 是最好的”，从而在训练结束时自动加载/保存那个最优 checkpoint，而不是默认用最后一个或用某个不符合业务目标的指标。

### 看懂训练过程

> {'loss': 0.0374, 'grad_norm': 403866.9375, 'learning_rate': 8.155090765588004e-06, 'epoch': 4.26}

## 评估与部署

TBD...

[^1]: https://huggingface.co/spaces/HuggingFaceTB/smol-training-playbook#post-training-compass-why--what--how
